### PR TITLE
[Sema] Check nullness of captured type before use

### LINKED ISSUE #72198: Clang front-end crash when lambda-capturing result of invalid structured binding
clang front-end crashes when lambda-capturing result of invalid structured binding.

Steps To Reproduce:

```c++
// test.cpp
void crash() {
    const int &[a, b] = {0, 0};
    auto x = [b]() -> int {
        return b;
    };
}
```

```
$ clang++ -std=gnu++17 -c test.cpp
```

Results:
Diagnostic output correctly reflects that program is invalid but the front-end also crashes.
https://godbolt.org/z/EfTbh1zKP

### PR DESCRIPTION
**Pull request by ecnelises**

### CODE REVIEW

**Comment by zyn0217** [comment]
Could you please explain why the `CaptureType` might be null at the time when exercising `tryCaptureVariable`? I'm worried that doing so would hide a more significant issue.

File: clang/test/SemaCXX/lambda-invalid-capture.cpp [hunk 1]
```diff
@@ -23,3 +23,9 @@ int pr43080(int i) { // expected-note {{declared here}}
       i; // expected-error {{variable 'i' cannot be implicitly captured in a lambda with no capture-default specified}}
   }();
 }
+
+void pr72198() {
+  int [_, b] = {0, 0}; // expected-error{{decomposition declaration cannot be declared with type 'int'; declared type must be 'auto' or reference to 'auto'}} \
+                          expected-error{{excess elements in scalar initializer}}
+  [b]{}; // expected-warning{{expression result unused}}
```
Comment (tbaederr) [inline clang/test/SemaCXX/lambda-invalid-capture.cpp:30]:
```suggestion
  int [_, b] = {0, 0}; // expected-error {{decomposition declaration cannot be declared with type 'int'; declared type must be 'auto' or reference to 'auto'}} \
                          expected-error {{excess elements in scalar initializer}}
  [b]{}; // expected-warning {{expression result unused}}
```

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -19773,6 +19773,8 @@ bool Sema::tryCaptureVariable(
   // declcontext can either capture the variable or have already captured
   // the variable.
   CaptureType = Var->getType();
+  if (CaptureType.isNull())
```
Comment (shafik) [inline clang/lib/Sema/SemaExpr.cpp:19776]: I am not sure this is the right fix. I think by the time we are here we should have caught any errors. I made a comment in the bug report describing my alternative.

**Comment by ecnelises** [comment]
Abandon in favor of #72428

**Pull request closed without merging**